src/optimizer/group_expression.py

```diff
@@ -0,0 +1,81 @@
+# coding=utf-8
+# Copyright 2018-2020 EVA
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from src.optimizer.operators import Operator
+from src.optimizer.group import INVALID_GROUP_ID
+from typing import List
+
+
+class GroupExpression:
+    def __init__(self,
+                 opr: Operator,
+                 group_id: int = INVALID_GROUP_ID,
+                 children: List[int] = []):
+        self._opr = opr
+        self._group_id = group_id
+        self._children = children
+        self._rules_explored = 0
+
+    @property
+    def opr(self):
+        return self._opr
+
+    @property
+    def group_id(self):
+        return self._group_id
+
+    @group_id.setter
+    def group_id(self, new_id):
+        self._group_id = new_id
+
+    @property
+    def children(self):
+        return self._children
+
+    @children.setter
+    def children(self, new_children):
+        self._children = new_children
+
+    def append_child(self, child_id: int):
+        self._children.append(child_id)
+
+    @property
+    def rules_explored(self):
+        return self._rules_explored
+
+    def mark_rule_explored(self, rule_id: int):
```

Comment:
@xzdandy @gaurav274 
I am little confused why we use bitwise OR here.
Let's say rule id 1 and id 2 are explored. Then `self._rule_explored = 3`. 
Then when we check rule id 3, it will mark as if rule id 3 has been explored.

Reply:
Actually, that's good point. I think rule id needs to be 1,2,4,8,16 to make this work.  So 
```
self._rules_explored |= ( 1<< rule_id)
self._rules_explored & (1 << rule_id)
```

Reply:
Yeah, it has to be 2's power. Then we can only have max of 64 rules in this case. It is good for now, but will it become an issue in the future if number of rules increases?

Reply:
Python int is not max at 64 bits. I tried `a = 1 << 1024` in python, and it works fine, though I do not know what is the max left bit shift operation Python can support.

Reply:
Wow, in Python3, integer can actually go infinite [source](https://stackoverflow.com/questions/7604966/maximum-and-minimum-values-for-ints).